clippy_lints/src/use_self.rs

```diff
@@ -218,18 +224,44 @@ struct UseSelfVisitor<'a, 'tcx> {
     cx: &'a LateContext<'a, 'tcx>,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for UseSelfVisitor<'a, 'tcx> {
-    fn visit_path(&mut self, path: &'tcx Path, _id: HirId) {
-        if path.segments.last().expect(SEGMENTS_MSG).ident.name != kw::SelfUpper {
-            if self.item_path.res == path.res {
-                span_use_self_lint(self.cx, path);
-            } else if let Res::Def(DefKind::Ctor(def::CtorOf::Struct, CtorKind::Fn), ctor_did) = path.res {
-                if self.item_path.res.opt_def_id() == self.cx.tcx.parent(ctor_did) {
-                    span_use_self_lint(self.cx, path);
+impl<'a, 'tcx: 'a> UseSelfVisitor<'a, 'tcx> {
+    fn check_path(&self, path: &hir::QPath) {
+        if let hir::QPath::Resolved(_, path) = path {
+            let last_segment = path.segments.last().expect(SEGMENTS_MSG);
+            if last_segment.ident.name != kw::SelfUpper {
+                if self.item_path.res == path.res {
+                    if SpanlessEq::new(&self.cx).eq_path(self.item_path, path) {
+                        span_use_self_lint(self.cx, path);
+                    }
+                } else if let Res::Def(DefKind::Ctor(def::CtorOf::Struct, CtorKind::Fn), ctor_did) = path.res {
+                    if self.item_path.res.opt_def_id() == self.cx.tcx.parent(ctor_did) {
+                        span_use_self_lint(self.cx, path);
+                    }
                 }
             }
         }
-        walk_path(self, path);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UseSelfVisitor<'a, 'tcx> {
+    fn visit_expr(&mut self, expr: &'tcx Expr) {
+        match &expr.node {
+            ExprKind::Struct(path, ..) => self.check_path(&*path),
+            ExprKind::Call(path, ..) => {
+                if let ExprKind::Path(path) = &path.node {
```

Comment:
Could you add a comment here why it is necessary to only `check_path`s if the path is from a call and therefore it is not possible to use `visit_path`?